src/validation.cpp

```diff
@@ -2680,22 +2672,30 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         if (!tx.IsCoinBase() && fScriptChecks)
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
-            bool tx_ok;
             TxValidationState tx_state;
-            // If CheckInputScripts is called with a pointer to a checks vector, the resulting checks are appended to it. In that case
-            // they need to be added to control which runs them asynchronously. Otherwise, CheckInputScripts runs the checks before returning.
-            if (control) {
-                std::vector<CScriptCheck> vChecks;
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
-                if (tx_ok) control->Add(std::move(vChecks));
-            } else {
-                tx_ok = CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
+            // If parallel script checking is possible (worker threads are available), they are added to control
+            // which runs the checks asynchronously. Otherwise they are run here directly.
+            std::vector<CScriptCheck> vChecks{};
+            ValidationCache& validation_cache{m_chainman.m_validation_cache};
+            if (PreCheckInputScripts(tx, view, flags, fCacheResults, txsdata[i], validation_cache)) {
+                // Turn all of this transaction's input scripts into script checks and add them to vChecks
+                vChecks.reserve(tx.vin.size());
+                for (unsigned int j = 0; j < tx.vin.size(); j++) {
+                    PrecomputedTransactionData& txdata{txsdata[i]};
+                    CScriptCheck check(txdata.m_spent_outputs[j], tx, validation_cache.m_signature_cache, j, flags, fCacheResults, &txdata);
+                    vChecks.emplace_back(std::move(check));
+                }
             }
-            if (!tx_ok) {
-                // Any transaction validation failure in ConnectBlock is a block consensus failure
-                state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
-                              tx_state.GetRejectReason(), tx_state.GetDebugMessage());
-                break;
+
+            if (control) control->Add(std::move(vChecks));
```

Comment:
Can you add braces here? I find different inlining styles for the same if-else block slightly confusing to read.